File: public/code/memecoin__social_momentum.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


# Strategy Lab - Social Mention Velocity
def checkSignal(data, cfg={}):
    minM = cfg.min_mentions_now  or  20
    minVel = cfg.min_velocity_pct  or  50
    minSent = cfg.min_sentiment  or  0.2
    minAuth = cfg.min_unique_authors  or  10
    _d = data
    symbol = _d['symbol']
    price = _d['price_usd']
    mN = _d['mentions_now']
    m1 = _d['mentions_1h_ago']
    m6 = _d['mentions_6h_ago']
    sent = _d['sentiment_score']
    kol = _d['kol_mentions']
    authors = _d['unique_authors_1h']
    if mN < minM:
        return make_signal("neutral", price, 0, [], 0, f"Low mentions ({mN})")
    velocity = ((mN - m1) / max(m1, 1)) * 100
    if velocity < minVel:
        return make_signal("neutral", price, 0, [], 0, f"Weak velocity ({velocity}%/h)")
    if sent < minSent:
        return make_signal("neutral", price, 0, [], 0, f"Weak sentiment ({sent})")
    if authors < minAuth:
        return make_signal("neutral", price, 0, [], 0, f"Few authors ({authors})")
    conf = 0.5
    if velocity > 200:
        conf += 0.15
    if sent > 0.6:
        conf += 0.1
    if kol >= 1:
        conf += 0.15
    if authors > 50:
        conf += 0.1
    conf = min(conf, 0.95)
    return make_signal("long", price, price*0.7, [price*1.5, price*2.5, price*4.0], conf, f"Social: {mN} mentions (+{velocity}%/h), sent={sent}, KOL={kol}", {'symbol': symbol, 'velocity_pct': velocity})

File: public/code/memecoin__social_momentum.ts
// Strategy Lab - Social Mention Velocity
import { makeSignal, Signal } from "../../common";

export interface SocialData {
  symbol: string;
  price_usd: number;
  mentions_now: number;
  mentions_1h_ago: number;
  mentions_6h_ago: number;
  sentiment_score: number;
  kol_mentions: number;
  unique_authors_1h: number;
}

export interface SocialConfig {
  min_mentions_now?: number;
  min_velocity_pct?: number;
  min_sentiment?: number;
  min_unique_authors?: number;
}

export function checkSignal(data: SocialData, cfg: SocialConfig = {}): Signal {
  const minM     = cfg.min_mentions_now ?? 20;
  const minVel   = cfg.min_velocity_pct ?? 50;
  const minSent  = cfg.min_sentiment ?? 0.2;
  const minAuth  = cfg.min_unique_authors ?? 10;

  const { symbol, price_usd: price, mentions_now: mN, mentions_1h_ago: m1, mentions_6h_ago: m6,
          sentiment_score: sent, kol_mentions: kol, unique_authors_1h: authors } = data;

  if (mN < minM) return makeSignal("neutral", price, 0, [], 0, `Low mentions (${mN})`);

  const velocity = ((mN - m1) / Math.max(m1, 1)) * 100;
  if (velocity < minVel)   return makeSignal("neutral", price, 0, [], 0, `Weak velocity (${velocity.toFixed(0)}%/h)`);
  if (sent < minSent)      return makeSignal("neutral", price, 0, [], 0, `Weak sentiment (${sent.toFixed(2)})`);
  if (authors < minAuth)   return makeSignal("neutral", price, 0, [], 0, `Few authors (${authors})`);

  let conf = 0.5;
  if (velocity > 200) conf += 0.15;
  if (sent > 0.6)     conf += 0.1;
  if (kol >= 1)       conf += 0.15;
  if (authors > 50)   conf += 0.1;
  conf = Math.min(conf, 0.95);

  return makeSignal("long", price, price*0.7, [price*1.5, price*2.5, price*4.0], conf,
                    `Social: ${mN} mentions (+${velocity.toFixed(0)}%/h), sent=${sent.toFixed(2)}, KOL=${kol}`,
                    { symbol, velocity_pct: velocity });
}
